### PR TITLE
Lazy-loading proxies package

### PR DESCRIPTION
**Pull request by ajcvickers**
Part of #10509, #3797.

This change adds a new package--Microsoft.EntityFrameworkCore.Proxies--that contains a lazy-loading proxy implementation making use of the EF lazy-loading infrastructure and the Castle.Core proxies package. Current plan is for this to ship on NuGet as an optional package for use with EF.

To use in a normal application, just add a call to `UseLazyLoadingProxies()` on the DbContext options builder.

There are also `context.CreateProxy()` methods for creating stand-alone proxy instances if needed.

Entity types must be public and navigation properties must be virtual. Also, the entity type and constructor must be "visible" to the castle proxies assembly, which usually means public, but could mean internal if "InternalsVisibleTo" is used. Exceptions are thrown if these requirements are not met--you don't just not get a proxy like in EF6.

Note that this is an optional package for EF that we chose to create because the infrastructure in place made it easy to do so. It does not preclude a Roslyn-based rewriting solution in the future.

### CODE REVIEW

**Comment by anpete** [comment]
Should this be called "CastleProxies"? Could it support different proxy impls? Or would they be separate packages?

**Comment by ajcvickers** [comment]
@anpete It can and probably will support different types of proxies. Theoretically those could use something other than Castle, but that's not the current plan. I'm hesitant to put a third-party reference in the package name.

@divega Thoughts?

File: src/EFCore/Metadata/Internal/PropertyParameterBinding.cs [hunk 1]
```diff
@@ -26,10 +27,14 @@ public PropertyParameterBinding([NotNull] IProperty consumedProperty)
         ///     directly from your code. This API may change or be removed in future releases.
         /// </summary>
         public override Expression BindToParameter(ParameterBindingInfo bindingInfo)
-            => Expression.Call(
-                EntityMaterializerSource.TryReadValueMethod.MakeGenericMethod(ConsumedProperty.ClrType),
+        {
+            var property = ConsumedProperties.First();
```
Comment (anpete) [inline src/EFCore/Metadata/Internal/PropertyParameterBinding.cs]: Save alloc with [0]?

File: src/EFCore.Proxies/Proxies/Internal/ProxyFactory.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+    ///     directly from your code. This API may change or be removed in future releases.
+    /// </summary>
+    public class ProxyFactory
+    {
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public static readonly ProxyFactory Instance = new ProxyFactory();
```
Comment (anpete) [inline src/EFCore.Proxies/Proxies/Internal/ProxyFactory.cs]: DI?

File: src/EFCore.Proxies/Proxies/Internal/ProxyFactory.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+        private readonly ProxyGenerator _generator = new ProxyGenerator();
```
Comment (anpete) [inline src/EFCore.Proxies/Proxies/Internal/ProxyFactory.cs]: DI?

File: src/EFCore/Metadata/Internal/PropertyParameterBinding.cs (hunk 1 above)
Comment (ajcvickers) [reply src/EFCore/Metadata/Internal/PropertyParameterBinding.cs]: Done

File: src/EFCore.Proxies/Proxies/Internal/ProxyFactory.cs (hunk 2 above)
Comment (ajcvickers) [reply src/EFCore.Proxies/Proxies/Internal/ProxyFactory.cs]: D.I. enabled the factory

**Comment by ralmsdeveloper** [comment]
@ajcvickers ,
I have a simple question.
Does using AsNotracking really override LazyLoad?

I ask why you when using: `UseLazyLoadingProxies();`

and use the AsNoTracking Ex: `db.Developers.AsNoTracking().ToList ();`

LazyLoad is simply ignored.
Can you understand my question?

**Comment by ajcvickers** [comment]
@ralmsdeveloper Loading of related entities (with .Load) does not work for entities that are not being tracked: see #10042. Since lazy loading is just a way of calling .Load it also doesn't support this. It's not even clear that this should work--I believe for 2.1 we will just make sure that it throws. In the future it could be made to work, but I doubt it will make it into 2.1.

**Comment by ralmsdeveloper** [comment]
Thanks @ajcvickers ,
I was breaking my head here, until i realized it.
I think there will be many people who need this, but great work for now!

Congratulations, this will eliminate some bad reviews!

**Pull request merged**